modes/play/objects/mob.js

```diff
@@ -0,0 +1,231 @@
+      this.recordCollision(pt[0], {x: x, y: y}, obj);
+      while (timeout > 0 && inRange(x, xRange) && inRange(y, yRange)) {
+        while (yTimeout > 0 && inRange(y, yRange)) {
+          vectY *= 0.9;
+          if (Math.abs(vectY) < 1) { vectY = 0 }
+          y = point.y + vectY;
+          yTimeout -= 1;
+        }
+
+        while (xTimeout && inRange(y, yRange)) {
+          vectX *= 0.9;
+          if (Math.abs(vectX) < 1) { vectX = 0 }
+          x = point.x + vectX;
+          xTimeout -= 1;
+        }
+        timeout -= 1
+        if (timeout < 1) { badNews = true };
+      };
+      if (badNews) {
+        timeout = 100;
+        console.log('Attempting to avert crisis');
+        let invX = this.velocity.x * -1;
+        let invY = this.velocity.y * -1;
+        x = point.x + invX;
+        y = point.y + invY;
+        while (timeout > 0 && inRange(x, xRange) && inRange(y, yRange)) {
+          invX *= 1.1
+          invY *= 1.1
+          timeout -= 1;
+        }
+        [vectX, vectY] = [invX, invY];
+      }
+      xVects.push(vectX);
+      yVects.push(vectY);
+    });
+    const finalVectX = xVects.reduce((a, b) => Math.abs(a) < Math.abs(b) ? a : b);
+    const finalVectY = yVects.reduce((a, b) => Math.abs(a) < Math.abs(b) ? a : b);
+    return [finalVectX, finalVectY];
+  }
+
+  recordCollision(name, coords, obj) {
+    const [xRange, yRange] = getRanges(obj);
+    if (inRange(coords.x, xRange) && inRange(coords.y, yRange)) {
+      if (this.collisionRecord[name]) {
+        this.collisionRecord[name].push(obj);
+      } else {
+        this.collisionRecord[name] = [obj];
+      }
+    };
+  }
+
+  physics() {
+    this.brain.think();
+    this.processCooldowns();
+    this.friction();
+    this.gravity();
+    this.processActions();
+    this.collisions([...this.mode.level.ground, ...this.mode.state.mobs]);
+    this.x += this.velocity.x
+    this.y += this.velocity.y;
+  }
+
+  gravity() {
+    if (this.velocity.y < this.fallSpeed) {
+      this.velocity.y = accelTo(this.velocity.y, this.fallSpeed, 2);
+    };
+  }
+
+  friction() {
+    if (this.velocity.x > -1 && this.velocity.x < 1) {
+      this.velocity.x = 0;
+    } else {
+      this.velocity.x *= this.frictionCoeff;
+    }
+  }
+
+  processActions() {
+    this.brain.actions.forEach(action => {
+      this.processAction(action);
+    });
+  }
+
+  processAction(action) {
+    if (action === 'left') {
+      return this.moveLeft();
+    } else if (action === 'right') {
+      return this.moveRight();
+    } else if (action === 'jump') {
+      this.jump();
+    } else if (action === 'run') {
+      this.run();
+    }
+  }
+
+  processCooldowns() {
+    if (Object.entries(this.cooldowns).length > 0) {
+      Object.entries(this.cooldowns).forEach(entry => {
+        const [name, time] = entry
+        this.cooldowns[name] -= 1
+        if (this.cooldowns[name] < 0) { this.cooldowns[name] = null }
+      });
+    }
+  }
+
+  setCooldown(name, time) {
```

Comment:
You probably sometimes want to be able to reset the timer when a thing happens. "Every time you get hit by a Stun Thing, you're stunned for 500ms." You're not immune to being stunned when you're already stunned, so the timer resets to 500ms every time. That's the default behavior here.

In other scenarios, you want the cooldown to set once, and then you can't touch it until it's expired... For example, "you can shoot once every 500ms." Regardless of how many times you hit the shoot button, you'll get a bullet every 500ms.

In yet other scenarios, maybe you want to EXTEND the timer. "Every time you collect a Power Thing, you get an additional 500ms to use your Super Move."

I just did the straightforward thing here... But I think cooldown timers are going to be really handy in the future.